torch/utils/_python_dispatch.py

```diff
@@ -164,3 +164,130 @@ def transform_subclass(t, callback):
+            assert hasattr(x, '__torch_dispatch__')
+
+    def alias_storage(arg, ret):
+        # This is hopefully a reasonable assert:
+        # subclasses that rely on this API for output aliasing
+        # should always return wrapper tensor subclasses for us to manually alias.
+        # in theory if a subclass that needs this API wants to sometimes return
+        # plain tensors, we could remove the assert and just not perform the aliasing,
+        # but it seems safer to learn more about this case first.
+        if is_traceable_wrapper_subclass(arg) or is_traceable_wrapper_subclass(ret):
+            assert type(arg) == type(ret), f"""Called {str(func)} with input of type {type(arg)}
+and output of type {type(ret)}. But expected types to match."""
+        # Need to run under no_dispatch, because we explicitly do **not**
+        # want our subclass to intercept the set_() call.
+        # instead, our subclass should directly have its storage swapped out.
+        with torch.utils._mode_utils.no_dispatch():
+            # directly calling this overload, and passing ret.shape, because we **explicitly**
+            # don't want to reset the sizes on ret, if the storage implies a size change.
+            # Why?
+            # The purpose of this API is *not* to change the size/strides of our output- we assume it's already correct.
+            # We just want to "fix up" the storage aliasing, without modifying or output's metadata.
+            # Example: out = inp.expand(inp.shape[0], inp.shape[0])
+            #     This requires swapping the storage of out to be the same as inp,
+            #     but we do *not* want it to change the sizes/strides that were compute for out.
+            meta_in_tls = torch._C._meta_in_tls_dispatch_include()
+            # TODO: make this a context manager
+            torch._C._set_meta_in_tls_dispatch_include(True)
+            try:
+                torch.ops.aten.set_.source_Storage_storage_offset(ret, arg.untyped_storage(), 0, ret.shape)
+            finally:
+                torch._C._set_meta_in_tls_dispatch_include(meta_in_tls)
+
+    def is_read_only_alias_match(arg, ret):
+        arg_aliases = set() if not arg.alias_info else arg.alias_info.before_set
+        out_aliases = set() if not ret.alias_info else ret.alias_info.before_set
+        shared_aliases = arg_aliases & out_aliases
+        return len(shared_aliases) > 0 and not arg.alias_info.is_write
+
+    num_args = len(func._schema.arguments)
+    num_returns = len(func._schema.returns)
+    for arg_idx in range(num_args):
+        for return_idx in range(num_returns):
+            if is_read_only_alias_match(func._schema.arguments[arg_idx], func._schema.returns[return_idx]):
+                alias_storage(args[arg_idx], outs[return_idx])
+
+    # Sigh... the torchscript parser has a bug where alias annotations for Tensor[](a) don't show up properly
+    # See https://github.com/pytorch/pytorch/issues/106173
```

Comment:
IIRC, when @larryliu0820 ran into these problems, we opted to use the torchgen schema parser instead, re-parsing the schema from Python, which seemed to work a lot better.

Reply:
Hmm, I could try that out (although tiny risk that using torchgen schema parsing on every op call through functionalization is a liiitle slow, maybe I should just do it for ops returning TensorList?)